src/test/java/org/jenkinsci/plugins/workflow/flow/StashManagerTest.java

```diff
@@ -0,0 +1,95 @@
+import static org.junit.Assert.assertTrue;
+
+public class StashManagerTest {
+    @ClassRule
+    public static BuildWatcher buildWatcher = new BuildWatcher();
+    @Rule
+    public RestartableJenkinsRule rr = new RestartableJenkinsRule();
+
+    @Issue("JENKINS-40912")
+    @Test
+    public void stashFileList() throws Exception {
+        rr.addStep(new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                WorkflowJob p = rr.j.jenkins.createProject(WorkflowJob.class, "p");
+                p.setDefinition(new CpsFlowDefinition(
+                        "node {\n" +
+                                "  dir('first') {\n" +
+                                "    writeFile file: 'fname', text: 'whatever'\n" +
+                                "    writeFile file: 'other', text: 'more'\n" +
+                                "  }\n" +
+                                "  dir('second') {\n" +
+                                "    writeFile file: 'at-top', text: 'ignored'\n" +
+                                "    dir('elsewhere') {\n" +
+                                "      writeFile file: 'nested', text: 'present'\n" +
+                                "      writeFile file: 'other', text: 'ignored'\n" +
+                                "    }\n" +
+                                "  }\n" +
+                                "}", true));
+                WorkflowRun b = rr.j.buildAndAssertSuccess(p);
+
+                // Not using stash step so that we're not dependent on a proper match between workflow-api and workflow-basic-steps versions.
+                FlowNode node = new DepthFirstScanner().findFirstMatch(b.getExecution(), new HasWorkspacePredicate());
+                assertNotNull(node);
+                WorkspaceAction action = node.getPersistentAction(WorkspaceAction.class);
+                assertNotNull(action);
+                FilePath ws = action.getWorkspace();
+                assertNotNull(ws);
+
+                TaskListener tl = rr.j.createTaskListener();
+
+                // Add stashes
+                StashManager.stash(b, "first", ws.child("first"), tl, null, null, true, false);
+                StashManager.stash(b, "second", ws.child("second"), tl, "elsewhere/", "**/other", true, false);
+
+                List<String> firstList = StashManager.stashFileList(b, "first");
+                assertFalse(firstList.isEmpty());
+                assertEquals(2, firstList.size());
+                assertTrue(firstList.contains("fname"));
+                assertTrue(firstList.contains("other"));
+
+                List<String> secondList = StashManager.stashFileList(b, "second");
+                assertFalse(secondList.isEmpty());
+                assertEquals(1, secondList.size());
+                assertTrue(secondList.contains("elsewhere/nested"));
+            }
+        });
+    }
+
+    private static final class HasWorkspacePredicate implements Predicate<FlowNode> {
```

Comment:
There is a utility for this in `workflow-api`.

Reply:
Yup, `FlowScanningUtils.hasActionPredicate(Class<? extends Action)` IIRC  🐜